torch/csrc/jit/autodiff.cpp

```diff
@@ -59,4 +57,170 @@ void differentiate(std::shared_ptr<Graph>& graph) {
   }
 }
 
+static std::shared_ptr<Graph> splitOffStage(
+        const std::shared_ptr<Graph>& graph,
+        std::size_t stage,
+        ArrayRef<Value*> inputs,
+        ArrayRef<Value*> outputs) {
+  auto graph_clone = std::make_shared<Graph>();
+
+  std::unordered_map<Value*, Value*> val_map; // values in graph -> values in graph_clone
+  const auto lookup_val = [&](Value *v) { return val_map.at(v); };
+
+  for (Value *input : inputs)
+    val_map[input] = graph_clone->addInput()->setType(input->typeOption());
+
+  for (Node *node : *graph) {
+    if (node->stage() != stage) continue;
+    Node *node_clone = graph_clone->createClone(node, lookup_val);
+    for (std::size_t i = 0, num_outputs = node_clone->outputs().size(); i < num_outputs; ++i)
+      val_map[node->outputs()[i]] = node_clone->outputs()[i];
+    graph_clone->appendNode(node_clone);
+  }
+
+  for (Value *output : outputs) {
+    JIT_ASSERT(output->stage() == stage);
+    graph_clone->registerOutput(val_map.at(output));
+  }
+
+  return graph_clone;
+}
+
+static std::unordered_map<Value*, int64_t> computeValueTopoOrder(const std::shared_ptr<Graph>& graph) {
+  int64_t next_nr = 0;
+  std::size_t stage = 0;
+  std::unordered_map<Value*, int64_t> order;
+  auto inputs_it = graph->inputs().begin();
+  auto inputs_end = graph->inputs().end();
+  auto nodes_it = graph->nodes().begin();
+  auto nodes_end = graph->nodes().end();
+
+  while (inputs_it != inputs_end || nodes_it != nodes_end) {
+    for (; inputs_it != inputs_end; ++inputs_it) {
+      Value *input = *inputs_it;
+      if (input->stage() != stage) break;
+      order[input] = next_nr++;
+    }
+    for (; nodes_it != nodes_end; ++nodes_it) {
+      Node *node = *nodes_it;
+      if (node->stage() != stage) break;
+      for (Value * output : node->outputs())
+        order[output] = next_nr++;
+    }
+    // NOTE: outputs are either inputs or are produced by nodes,
+    // so we don't need to handle them in any way.
+    stage++;
+  }
+  JIT_ASSERT(stage == graph->stage() + 1);
+  return order;
+}
+
+LiftedReverse lambdaLiftReverse(std::shared_ptr<Graph>& graph) {
+  static const auto is_stage_0 = [](Value *v) { return v->stage() == 0; };
+  static const auto is_stage_1 = [](Value *v) { return v->stage() == 1; };
+  // NOTE: in the comments here first stage == stage 0
+  JIT_ASSERT(graph->stage() == 1);
+
+  // --------------------------------------------------------------------------
+  // 1. Find values of stage 0 that need to ba captured.
+  // --------------------------------------------------------------------------
+  // First, we need to find all values that are produced in the first stage,
+  // and used in the second one. They will need to be added as inputs of the reverse
+  // graph, and some of them may also need to be appended as outputs of the primal graph.
+  std::unordered_set<Value*> extra_reverse_inputs;
```

Comment:
Do this the opposite way: look for definitions of stage 0 values used in stage one. This will naturally find them in topological order, no need for a separate pass.